message BFT2F_VERSION {
  required int64 node_id = 1;
  required int64 view = 2;
  required int64 n = 3;
  optional string hcd = 4;
  optional string sig = 5;
}

enum BFT2F_OP_TYPE {
  SIGN_UP = 0;
  SIGN_IN = 1;
  CHANGE_CRED = 2;
  NO_OP = 3;
}

message BFT2F_OP {
  required BFT2F_OP_TYPE type = 1;
  required string user_id = 2;
  optional string token = 3;
  optional string user_pub_key = 4;
  optional string user_priv_key_enc = 5;
  optional string new_user_pub_key = 6;
  optional string new_user_priv_key_enc = 7;
  optional string sig = 8;
}

message BFT2f_OP_RES {
  enum BFT2f_OP_RES_TYPE {
    SUCCESS = 0;
    USER_ID_NOT_FOUND = 1;
    USER_ID_EXISTS = 2;
    INVALID_CRED_CHANGE = 3;
  }
  required BFT2f_OP_RES_TYPE type = 1;
  required BFT2F_OP_TYPE op_type = 2;
  required string user_id = 3;
  optional string user_pub_key = 4;
  optional string user_priv_key_enc = 5;
  optional string token = 6;
  optional BFT2f_SIGN_IN_CERT sign_in_cert = 7;
}

message BFT2f_SIGN_IN_CERT {
  required string node_pub_key = 1;
  required string sig = 2;
}

message BFT2F_MESSAGE {
  enum BFT2F_MESSAGE_TYPE {
    REQUEST = 0;
    PRE_PREPARE = 1;
    PREPARE = 2;
    COMMIT = 3;
    REPLY = 4;
    VIEW_CHANGE = 5;
    NEW_VIEW = 6;    
    CHECKPOINT = 7;
    FAST_FORWARD_REQUEST = 8;
    FAST_FORWARD_REPLY = 9;
  }

  required BFT2F_MESSAGE_TYPE msg_type = 1;
  optional string sig = 2;
  optional BFT2F_OP op = 3;
  optional int64 ts = 4;
  optional int64 client_id = 5;
  optional BFT2F_VERSION version = 6;
  optional int64 node_id = 7;
  optional int64 view = 8;
  optional int64 n = 9;
  optional string req_D = 10;
  repeated BFT2F_MESSAGE_LIST P = 11;
  repeated BFT2F_MESSAGE O = 12;
  repeated BFT2F_MESSAGE V = 13;
  optional BFT2f_OP_RES res = 14;
  repeated BFT2F_REQUEST_PROOF req_proofs = 15;
  optional string state_D = 16;
  optional string replay_cache_D = 17;
}

message BFT2F_MESSAGE_LIST {
  repeated BFT2F_MESSAGE msgs = 1;
}

message BFT2F_REQUEST_PROOF {
  required BFT2F_MESSAGE req = 1;
  repeated BFT2F_VERSION matching_versions = 2;
}

